enzy_htp/core/general.py

```diff
@@ -0,0 +1,133 @@
+"""This module contains helper functions related to general python/python library features
+
+Author: QZ Shao, <shaoqz@icloud.com>
+Date: 2022-10-21
+"""
+import copy
+import numpy as np
+from typing import List, Iterable, Tuple, Dict
+from time import perf_counter
+import itertools
+
+from .logger import _LOGGER
+
+
+# == List related ==
+class GhostListElement:
+    _instance = None
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = object.__new__(cls)
+            cls._instance._value = cls._instance
+        return cls._instance
+
+    def __eq__(self, other):
+        return self is other
+
+
+GHOST_LIST_ELEMENT = GhostListElement()
+"""singleton for a ghost list element used for product_lists_allow_empty"""
+
+
+def delete_base_on_id(target_list: list, target_id: int):
+    """
+    delete an element from a list base on its id() value
+    """
+    for i in range(len(target_list) - 1, -1, -1):
+        if id(target_list[i]) == target_id:
+            del target_list[i]
+
+
+def get_interval_from_list(target_list: List[int]) -> Iterable[Tuple[int, int]]:
+    """
+    convert a list of int to the interval/range representation
+    Returns:
+        a generater of tuples with each indicating the start/end of the interval
+    Example:
+        >>> list(get_interval_from_list([1,2,3,6,7,8]))
+        [(1,3),(6,8)]
+    reference: https://stackoverflow.com/questions/4628333
+    """
+    # clean input
+    target_list = sorted(set(target_list))
+    # here use enum id as a ref sequence and group by the deviation
+    for i, j in itertools.groupby(
+            enumerate(target_list),
+            lambda ref_vs_target: ref_vs_target[1] - ref_vs_target[0]):
+        j = list(j)
+        yield j[0][1], j[-1][1]
+
+
+def get_interval_str_from_list(target_list: List[int]) -> str:
+    """convert the result of get_interval_from_list() to a str"""
+    result = get_interval_from_list(target_list)
+    result = ",".join([f"{x[0]}-{x[1]}" if x[0] != x[1] else f"{x[0]}" for x in result])
+    return result
+
+
+def get_random_list_elem(traget_list: list):
+    """Helper method that randomly chooses an element from a list. numpy.random.choice() doesn't 
+    like to take elements from list()'s of tuples so this is the work around."""
+    return traget_list[np.random.randint(len(traget_list))]
```

Comment:
Use np.random.choice() and add a `seed` argument

Reply:
I think seeding is a global behavior. That's the reason I didn't add `seed` for all random related functions.
I will change the np function